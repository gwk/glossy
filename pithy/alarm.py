# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import signal
from typing import Callable


class Timeout(Exception):
  'Exception indictaing that an AlarmManager timed out.'


class AlarmManager:

  def __init__(self, timeout: int, msg:str='alarm timed out after {timeout} seconds', on_signal: Callable[[], None]=None) -> None:
    if not isinstance(timeout, int): raise TypeError(f'timeout must be an int; received: {timeout!r}')
    if timeout < 0: raise TypeError(f'timeout must be nonnegative; received: {timeout!r}')
    self.timeout = timeout
    self.msg = msg
    self.on_signal = on_signal
    self.timed_out = False


  def __enter__(self):
    if not self.timeout: return
    prev = signal.getsignal(signal.SIGALRM)
    if prev is not signal.SIG_DFL:
      raise Exception(f'task.run encountered previously installed signal handler: {prev}')

    def alarm_handler(signum, current_stack_frame):
      # since signal handlers carry reentrancy concerns; do not do any IO within the handler.
      self.timed_out = True
      if self.on_signal: self.on_signal() # callback must respect reentrancy limitations.

    signal.signal(signal.SIGALRM, alarm_handler) # set handler.
    signal.alarm(self.timeout) # set alarm.


  def __exit__(self, exc_type, exc_value, traceback) -> None:
    if not self.timeout: return None
    signal.alarm(0) # disable alarm.
    signal.signal(signal.SIGALRM, signal.SIG_DFL) # uninstall handler.
    if exc_type: return None # exception will be reraised.
    if self.timed_out: raise Timeout(self.msg.format(timeout=self.timeout))
    return None
