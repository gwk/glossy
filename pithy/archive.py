# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from functools import total_ordering
from io import BytesIO, TextIOBase, TextIOWrapper
from typing import Any, BinaryIO, Callable, Iterable, List, Tuple, Union
from tarfile import TarFile
from zipfile import ZipFile, BadZipFile
from .fs import path_ext
from .util import lazy_property
from .io import *


FileOrPath = Union[BinaryIO, str] # TODO: support other pathlike objects.
Opener = Callable[[], BinaryIO]


class Archive:

  def __init__(self, file_or_path:FileOrPath, ext=None) -> None:
    self.name: str = getattr(file_or_path, 'name', str(file_or_path))
    if ext is None:
      ext = path_ext(self.name)
    self.ext = ext
    try: self._handler = _ext_handlers[ext](file_or_path)
    except KeyError: raise ValueError(f'Archive does not support extension: {ext!r}')


  def __repr__(self) -> str:
    return f'{self.__class__.__name__}(name={self.name!r})'


  def __iter__(self) -> Iterable['ArchiveFile']:
    return self._handler.files(self)



@total_ordering
class ArchiveMember:

  def __init__(self, archive: Archive, name: str) -> None:
    self.archive = archive
    self.name = name

  def __repr__(self) -> str:
    return f'{self.__class__.__name__}({self.archive}, {self.name!r})'

  def __lt__(self, other: Any) -> bool:
    if isinstance(other, ArchiveMember):
      return self.name < other.name
    return NotImplemented



class ArchiveFile(ArchiveMember):

  def __init__(self, archive:Archive, name:str, opener:Opener) -> None:
    super().__init__(archive=archive, name=name)
    self._opener = opener

  @lazy_property
  def _file(self): return self._opener()

  def __iter__(self): return iter(self._file)

  def __getattr__(self, name):
    # note: dunder attribute access is never forwarded to this method.
    if name.startswith('_'): raise AttributeError(name)
    return getattr(self._file, name)

  def seekable(self):
    return getattr(self._file, 'seekable', False)

  def text(self, encoding='UTF-8', errors=None, newline=None) -> TextIOWrapper:
    'Returns a TextIOWrapper around the underlying binary buffer.'
    return TextIOWrapper(self._file, encoding=encoding, errors=errors, newline=newline)



class ArchiveDir(ArchiveMember): pass


class _Handler:
  def __init__(self, file_or_path:FileOrPath) -> None: raise NotImplementedError()


class _TarHandler(_Handler):
  def __init__(self, file_or_path:FileOrPath) -> None:
    self.tar = TarFile.open(mode='r|*', fileobj=file_or_path)

  def files(self, archive:Archive) -> Iterable[Tuple[str, Opener]]:
    for info in self.tar:
      if info.isdir(): continue
      opener = lambda: self.tar.extractfile(info)
      #^ note: crucial to pass info here and not name, or else gz archives get very slow.
      yield ArchiveFile(archive=archive, name=info.name, opener=opener)


class _ZipHandler(_Handler):
  def __init__(self, file_or_path:FileOrPath) -> None:
    try: self.zip = ZipFile(file_or_path)
    except BadZipFile as e:
      if isinstance(file_or_path, TextIOBase):
        raise TypeError(f'Archive requires a path or binary stream but received text stream: {file_or_path}') from e
      raise

  def files(self, archive:Archive) -> Iterable[Tuple[str, Opener]]:
    for name in self.zip.namelist():
      if name.endswith('/'): continue # directory.
      opener = lambda: self.zip.open(name)
      yield ArchiveFile(archive=archive, name=name, opener=opener)


_ext_handlers = {
  '.bz2' : _TarHandler,
  '.gz'  : _TarHandler,
  '.tar' : _TarHandler,
  '.xz', : _TarHandler,
  '.zip' : _ZipHandler,
}
