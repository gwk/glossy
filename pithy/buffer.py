# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from typing import Callable, Generic, Iterable, Iterator, List, TypeVar, Union
from .default import Raise, RaiseOr


T = TypeVar('T')


class Buffer(Iterator[T]):
  '''
  Iterable object that buffers an iterable.
  Call push() to push an item into the buffer;
  this will be returned on the subsequent call to next().
  '''


  def __init__(self, iterable: Iterable[T]) -> None:
    self.iterator = iter(iterable)
    self.buffer: List[T] = []


  def __repr__(self) -> str:
    return 'IterBuffer({!r}, buffer={!r})'.format(self.iterator, self.buffer)


  def __iter__(self) -> Iterator[T]: return self


  def __next__(self) -> T:
    try: return self.buffer.pop()
    except IndexError: pass
    return next(self.iterator)


  @property
  def is_live(self) -> bool:
    try: self.peek()
    except StopIteration: return False
    else: return True


  def push(self, item: T) -> None:
    self.buffer.append(item)


  def peek(self, default: RaiseOr[T]=Raise._) -> T:
    try: return self.buffer[-1]
    except IndexError: pass
    try: el = next(self.iterator)
    except StopIteration:
      if isinstance(default, Raise): raise
      else: return default
    self.buffer.append(el)
    return el


  def take_while(self, pred: Callable[[T], bool]) -> Iterator[T]:
    for el in self:
      if pred(el):
        yield el
      else:
        self.buffer.append(el)
        break


  def drop_while(self, pred: Callable[[T], bool]) -> None:
    for el in self:
      if not pred(el):
        self.buffer.append(el)
        break


  def peek_while(self, pred: Callable[[T], bool]) -> List[T]:
    els = list(self.take_while(pred))
    self.buffer.extend(reversed(els))
    return els


  def take(self, count: int, short=False, default: RaiseOr[T]=Raise._) -> List[T]:
    els = []
    for _ in range(count):
      try: els.append(next(self))
      except StopIteration:
        if short: break
        if isinstance(default, Raise): raise
        els.append(default)
    return els


  def peeks(self, count: int, short=False, default: RaiseOr[T]=Raise._) -> List[T]:
    if 0 < count <= len(self.buffer):
      return list(reversed(self.buffer[-count:]))
    els = []
    for _ in range(count):
      try: els.append(next(self))
      except StopIteration:
        if short: break
        if isinstance(default, Raise): raise
        els.append(default)
    self.buffer.extend(reversed(els))
    return els


  def expect(self, pred: Callable[[T], bool]) -> T:
    el = next(self)
    if pred(el): return el
    raise ValueError(el)
