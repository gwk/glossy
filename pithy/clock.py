# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from sys import stderr
from time import time as now


class Clock:
  def __init__(self, name:str):
    self.name = name
    self.times = [(now(), 'start')]

  def split(self, label:str):
    t = now()
    print(f'clock: {self.name}: {label}: {t - self.times[-1][0]:.04f}', file=stderr)
    self.times.append((t, label))


  def total(self):
    t = now()
    print(f'clock: {self.name}: TOTAL: {t - self.times[0][0]:.04f}', file=stderr)
