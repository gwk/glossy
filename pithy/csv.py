# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

# TODO: handle leading UTF8-BOM.

import csv
import collections
from sys import stdout
from types import TracebackType
from typing import ContextManager, Iterable, Iterator, Optional, Sequence, TextIO, Type, Union


def load_csv(file: TextIO,
 dialect:Optional[str]=None,
 delimiter:Optional[str]=None,
 doublequote:Optional[bool]=None,
 escapechar:Optional[str]=None,
 quotechar:Optional[str]=None,
 quoting:Optional[int]=None,
 skipinitialspace:Optional[bool]=None,
 strict:Optional[bool]=None,
 header:Union[None, bool, Sequence[str]]=None) -> 'CSVFileReader':

  return CSVFileReader(
    file=file,
    dialect=dialect,
    delimiter=delimiter,
    doublequote=doublequote,
    escapechar=escapechar,
    quotechar=quotechar,
    quoting=quoting,
    skipinitialspace=skipinitialspace,
    strict=strict,
    header=header)


def write_csv(f:TextIO, *, header:Optional[Sequence[str]], rows:Iterable[Sequence]) -> None:
  w = csv.writer(f)
  if header is not None: w.writerow(header)
  w.writerows(rows)


def out_csv(*, header:Optional[Sequence[str]], rows:Iterable[Sequence]) -> None:
  write_csv(f=stdout, header=header, rows=rows)


class CSVFileReader(Iterable, ContextManager):

  def __init__(self, file: TextIO,
   dialect:Optional[str]=None,
   delimiter:Optional[str]=None,
   doublequote:Optional[bool]=None,
   escapechar:Optional[str]=None,
   quotechar:Optional[str]=None,
   quoting:Optional[int]=None,
   skipinitialspace:Optional[bool]=None,
   strict:Optional[bool]=None,
   header:Union[None, bool, Sequence[str]]=None) -> None:

    opts = { k : v for (k, v) in [
      ('dialect', dialect),
      ('delimiter', delimiter),
      ('doublequote', doublequote),
      ('escapechar', escapechar),
      ('quotechar', quotechar),
      ('quoting', quoting),
      ('skipinitialspace', skipinitialspace),
      ('strict', strict),
      ] if v is not None }

    self._reader = csv.reader(file, dialect, **opts)
    self.file = file

    if header is None or isinstance(header, bool):
      if header: next(self._reader) # simply discard.
    else: # match expected against actual.
      row = next(self._reader)
      list_header = list(header)
      if row != list_header:
        raise ValueError(f'load_csv expected header:\n{list_header}\nreceived:\n{row}')


  def __iter__(self) -> Iterator[Sequence[str]]:
    return self._reader

  def __enter__(self) -> 'CSVFileReader':
    return self

  def __exit__(self, exc_type: Optional[Type[BaseException]], exc_value: Optional[BaseException],
   traceback: Optional[TracebackType]) -> None:
    self.file.close()
