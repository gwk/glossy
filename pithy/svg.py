# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

'''
SVG writer.
SVG elements reference: https://developer.mozilla.org/en-US/docs/Web/SVG/Element.
'''

from .num import Num, NumRange
from .xml import XmlAttrs, XmlWriter, add_opt_attrs, esc_xml_attr, esc_xml_text
from html import escape as html_escape
from types import TracebackType
from typing import Any, ContextManager, Dict, Iterator, List, Optional, Sequence, TextIO, Tuple, Type, Union, Iterable


Dim = Union[int, float, str]
Vec = Tuple[Num, Num]
VecOrNum = Union[Vec, Num]
PathCommand = Tuple

ViewBox = Union[None, Vec, Tuple[Num, Num, Num, Num], Tuple[Vec, Vec]] # TODO: currently unused.


class SvgWriter(XmlWriter):
  '''
  SvgWriter is a ContextManager class that outputs SVG code to a file (stdout by default).
  Like its parent class XmlWriter, it uses the __enter__ and __exit__ methods to automatically output open and close tags.
  '''

  replaced_attrs = {
    'href': 'xlink:href', # safari Version 11.1.1 requires this, even though xlink is deprecated in svg 2 standard.
    **XmlWriter.replaced_attrs,
  }


  def __init__(self, file:TextIO=None, x:Dim=None, y:Dim=None, w:Dim=None, h:Dim=None,
   vx:Num=None, vy:Num=None, vw:Num=None, vh:Num=None, **attrs:Any) -> None:
    self.x = x
    self.y = y
    self.w = w
    self.h = h
    self.vx = vx
    self.vy = vy
    self.vw = vw
    self.vh = vh
    self.viewBox = fmt_viewBox(vx, vy, vw, vh)
    attrs = { # Put the xml nonsense up front.
      'xmlns': 'http://www.w3.org/2000/svg',
      'xmlns:xlink': 'http://www.w3.org/1999/xlink', # Safari does not yet support SVG 2.
      **attrs
    }
    add_opt_attrs(attrs, x=x, y=y, width=w, height=h, viewBox=self.viewBox)
    super().__init__(tag='svg', file=file, attrs=attrs)


  def leaf(self, tag:str, attrs:XmlAttrs) -> None:
    '''
    Output a non-nesting SVG element.
    'title' is treated as a special attribute that is translated into a <title> child element and rendered as a tooltip.
    '''
    if attrs and 'title' in attrs: self.leaf_text(tag, attrs, text='')
    else: super().leaf(tag, attrs)


  def leaf_text(self, tag:str, attrs:XmlAttrs, text:str) -> None:
    '''
    Output a non-nesting XML element that contains text between the open and close tags.
    'title' is treated as a special attribute that is translated into a <title> child element and rendered as a tooltip.
    '''
    if attrs:
      try: title = attrs.pop('title')
      except KeyError: pass
      else:
        self.write(f'<{tag}{self.fmt_attrs(attrs)}>{self.title(title)}{esc_xml_text(text)}</{tag}>')
        return
    super().leaf_text(tag, attrs=attrs, text=text)


  def sub(self, tag:str, attrs:XmlAttrs) -> XmlWriter:
    '''
    Create a child XmlWriter for use in a `with` context to represent a nesting XML element.
    'title' is treated as a special attribute that is translated into a <title> child element and rendered as a tooltip.
    '''
    title_el = ''
    if attrs:
      try: title = attrs.pop('title')
      except KeyError: pass
      else:
        title_el = self.title(title)
    s = super().sub(tag, attrs=attrs)
    if title_el: self.write(title_el)
    return s


  # SVG Elements.

  def title(self, title:Optional[str]) -> str:
    return '' if title is None else f'<title>{esc_xml_text(title)}</title>'


  def circle(self, pos:Vec=None, r:Num=None, *, x:Num=None, y:Num=None, **attrs) -> None:
    'Output an SVG `circle` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    add_opt_attrs(attrs, cx=x, cy=y, r=r)
    self.leaf('circle', attrs)


  def defs(self, **attrs:Any) -> XmlWriter:
    'Output an SVG `defs` element.'
    return self.sub('defs', attrs)


  def g(self, *transforms:str, **attrs:Any) -> XmlWriter:
    'Create an SVG `g` element for use in a context manager.'
    add_opt_attrs(attrs, transform=(' '.join(transforms) if transforms else None))
    return self.sub('g', attrs)


  def image(self, pos:Vec=None, size:VecOrNum=None, *, x:Num=None, y:Num=None, w:Num=None, h:Num=None, **attrs) -> None:
    'Output an SVG `defs` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    if size is not None:
      assert w is None
      assert h is None
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    add_opt_attrs(attrs, x=x, y=y, width=w, height=h)
    self.leaf('image', attrs)


  def line(self, a:Vec=None, b:Vec=None, *, x1:Num=None, y1:Num=None, x2:Num=None, y2:Num=None, **attrs) -> None:
    'Output an SVG `defs` element.'
    if a is not None:
      assert x1 is None
      assert y1 is None
      x1, y1 = a
    if b is not None:
      assert x2 is None
      assert y2 is None
      x2, y2 = b
    add_opt_attrs(attrs, x1=x1, y1=y1, x2=x2, y2=y2)
    self.leaf('line', attrs)


  def marker(self, id:str, pos:Vec=None, size:VecOrNum=None, *, x:Num=None, y:Num=None, w:Num=None, h:Num=None,
   vx:Num=None, vy:Num=None, vw:Num=None, vh:Num=None,
   markerUnits='strokeWidth', orient:str='auto', **attrs:Any) -> XmlWriter:
    'Output an SVG `marker` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    if size is not None:
      assert w is None
      assert h is None
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    add_opt_attrs(attrs, id=id, refX=x, refY=y, markerWidth=w, markerHeight=h, viewBox=fmt_viewBox(vx, vy, vw, vh),
      markerUnits=markerUnits, orient=orient)
    return self.sub('marker', attrs=attrs)


  def path(self, commands:Iterable[PathCommand], **attrs:Any) -> None:
    'Output an SVG `path` element.'
    assert 'd' not in attrs
    cmd_strs:List[str] = []
    for c in commands:
      try: code = c[0]
      except IndexError: continue
      try: exp_len = path_command_lens[code]
      except KeyError as e: raise Exception(f'bad path command code: {c!r}') from e
      if len(c) != exp_len + 1: raise Exception(f'path command requires {exp_len} arguments: {c}')
      cmd_strs.append(code + ','.join(str(x) for x in c[1:]))
    assert 'd' not in attrs
    attrs['d'] = ' '.join(cmd_strs)
    self.leaf('path', attrs)


  def polyline(self, points:Iterable[Vec], **attrs:Any) -> None:
    point_strs:List[str] = []
    assert 'points' not in attrs
    for p in points:
      if len(p) < 2: raise Exception(f'bad point for polyline: {p}')
      point_strs.append(f'{p[0]},{p[1]}')
    attrs['points'] = ' '.join(point_strs)
    self.leaf('polyline', attrs)


  def rect(self, pos:Vec=None, size:VecOrNum=None, *, x:Num=None, y:Num=None, w:Num=None, h:Num=None, r:VecOrNum=None, **attrs:Any) -> None:
    'Output an SVG `rect` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    if size is not None:
      assert w is None
      assert h is None
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    rx:Optional[Num]
    ry:Optional[Num]
    if isinstance(r, tuple):
      rx, ry = r
    else:
      rx = ry = r
    add_opt_attrs(attrs, x=x, y=y, width=w, height=h, rx=rx, ry=ry)
    self.leaf('rect', attrs)


  def style(self, text:str, **attrs,) -> None:
    'Output an SVG `style` element.'
    self.leaf_text('style', attrs, text.strip())


  def symbol(self, id:str, vx:Num=None, vy:Num=None, vw:Num=None, vh:Num=None, **attrs:Any) -> XmlWriter:
    'Output an SVG `symbol` element.'
    if vx is None: vx = 0
    if vy is None: vy = 0
    # TODO: figure out if no viewBox is legal and at all useful.
    assert vw >= 0 # type: ignore
    assert vh >= 0 # type: ignore
    add_opt_attrs(attrs, id=id, viewBox=f'{vx} {vy} {vw} {vh}')
    return self.sub('symbol', attrs)


  def text(self, pos:Vec=None, x:Num=None, y:Num=None, text=None, **attrs) -> None:
    'Output an SVG `text` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    add_opt_attrs(attrs, x=x, y=y)
    self.leaf_text('text', attrs, text)


  def use(self, id:str, pos:Vec=None, size:VecOrNum=None, x:Num=None, y:Num=None, w:Num=None, h:Num=None, **attrs) -> None:
    'Use a previously defined symbol'
    assert id
    if id[0] != '#': id = '#' + id
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    if size is not None:
      assert w is None
      assert h is None
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    add_opt_attrs(attrs, href=id, x=x, y=y, width=w, height=h)
    return self.leaf('use', attrs)


  # High level.

  def grid(self, pos:Vec=None, size:VecOrNum=None, *,
   x:Num=0, y:Num=0, w:Num=256, h:Num=256, step:VecOrNum=16, off:Vec=(0, 0), r:VecOrNum=None, **attrs:Any) -> None:
    # TODO: axis transformers (e.g. log-log).
    assert step is not None
    if isinstance(step, tuple):
      sx, sy = step
    else:
      sx = sy = step
    if pos is not None:
      x, y = pos
    if size is not None:
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    x = float(x)
    y = float(y)
    w = float(w)
    h = float(h)
    off_x, off_y = off
    if off_x <= 0: off_x = sx
    if off_y <= 0: off_y = sy
    x_start = x + off_x
    y_start = y + off_y
    x_end = x + w
    y_end = y + h
    class_ = attrs.pop('class_', 'grid')
    with self.g(class_=class_, **attrs):
      self.rect(class_=class_+'-border', x=x, y=y, w=w, h=h, r=r)
      for tick in NumRange(x_start, x_end, sx): self.line((tick, y), (tick, y_end)) # Vertical lines.
      for tick in NumRange(y_start, y_end, sy): self.line((x, tick), (x_end, tick)) # Horizontal lines.


def scale(x:Num=1, y:Num=1) -> str: return f'scale({x},{y})'
def rotate(degrees:Num, x:Num=0, y:Num=0) -> str: return f'rotate({degrees},{x},{y})'
def translate(x:Num=0, y:Num=0) -> str: return f'translate({x},{y})'


path_command_lens = {
  'A' : 7,
  'M' : 2,
  'm' : 2,
  'L' : 2,
  'l' : 2,
  'Z' : 0,
  'z' : 0
}

valid_units = frozenset({'em', 'ex', 'px', 'pt', 'pc', 'cm', 'mm', '%', ''})

def _validate_unit(unit: str):
  'Ensure that `unit` is a valid unit string.'
  if unit not in valid_units:
    raise Exception(f'Invalid SVG unit: {unit!r}; should be one of {sorted(valid_units)}')


def fmt_viewBox(vx:Optional[Num], vy:Optional[Num], vw:Optional[Num], vh:Optional[Num]) -> Optional[str]:
  if vx is None and vy is None and vw is None and vh is None:
    return None
  else:
    if vx is None: vx = 0
    if vy is None: vy = 0
    assert vw is not None and vw > 0
    assert vh is not None and vh > 0
    return f'{vx} {vy} {vw} {vh}'


baselines = {
  'after-edge',
  'alphabetic',
  'auto',
  'baseline',
  'before-edge',
  'central',
  'hanging',
  'ideographic',
  'inherit',
  'mathematical',
  'middle',
  'text-after-edge',
  'text-before-edge',
}
