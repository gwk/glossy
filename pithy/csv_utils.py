# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import csv


def write_csv(f, header, rows):
  w = csv.writer(f)
  w.writerow(header)
  w.writerows(rows)

