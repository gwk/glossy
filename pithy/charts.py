from .svg import *
from io import StringIO
from typing import *

class XYChart:
  def __init__(self, *series:Tuple[str, Iterable[Tuple]],
   title:str=None,
   x_label:str=None, y_label:str=None,
   w:int=None, h:int=None,
   plot_w:int=None, plot_h:int=None,
   x_min:int=None, x_max:int=None, y_min:int=None, y_max:int=None,
   title_height:int=None,
   label_width:int=None, label_height:int=None,
   legend_width:int=None, # TODO: legend_height would imply that the legend is at bottom.
   ) -> None:
    self.series = {k : list(v) for k, v in series}

    #take smallest_x and largest_x value from data if x_min or y_min is undefined
    min_x:float = 0
    max_x:float = 1
    min_y:float = 0
    max_y:float = 1
    for k,s in self.series.items():
      for x,y in s:
        min_x = min(x, min_x)
        max_x = max(x, max_x)
        min_y = min(y, min_y)
        max_y = max(y, max_y)
    if(x_min == None): x_min = min_x
    if(y_min == None): y_min = min_y
    if(x_max == None): x_max = max_x
    if(y_max == None): y_max = max_y
    self.x_min = x_min
    self.x_max = x_max
    self.y_min = y_min
    self.y_max = y_max
    x_range:float = x_max - x_min
    y_range:float = y_max - y_min
    self.svg_width = w
    self.svg_height = h
    self.label_height = label_height
    self.label_width = label_width
    self.legend_width = legend_width
    self.title_height = title_height
    self.title = title
    self.x_label = x_label
    self.y_label = y_label

    x_scale = ((self.svg_width - 10) - (self.legend_width + self.label_width + 10))/x_range
    y_scale = ((self.svg_height - self.title_height)-(self.label_height+10))/y_range
    scaled_min_x=0
    scaled_min_y=0
    for k,s in self.series.items():
      for x,y in s:
        scaled_min_x = min(x*x_scale, scaled_min_x)
        scaled_min_y = min(y*y_scale, scaled_min_y)
    for k,s in self.series.items():
      temp_list = []
      for x,y in s:
        temp_list.append(((x*x_scale)-scaled_min_x, (y*y_scale)-scaled_min_y))
      self.series[k] = temp_list



  def render(self, file=None) -> str:
    colors:List = ['#EF7F6D', '#F78398', '#E893C4', '#C2AAE5', '#89C2F3', '#49D5E9', '#38E2CA', '#70E9A0', '#AFEB77', '#EFE55D']
    if file is None:
      file = StringIO()
    chart_x_min:int = self.legend_width + self.label_width + 10
    chart_x_max:int = self.svg_width - 10
    chart_y_min:int = self.label_height+10
    chart_y_max:int = self.svg_height - self.title_height
    with SvgWriter(file, w=self.svg_width, h=self.svg_height) as svg:
      y_range = chart_y_max - chart_y_min
      #draw graph
      svg.rect(pos=(chart_x_min, chart_y_min), size=(chart_x_max-chart_x_min, chart_y_max - chart_y_min), stroke=None, fill='#fafafa')
      transform_x =  700
      color_picker = 0
      with svg.g(transform=f'translate({chart_x_min} {chart_y_max}) scale(1 -1) '):
        for i in range(1,5):
            svg.line((0, (y_range/5)*i), (chart_x_max-70, (y_range/5)*i), stroke='#BBCCDD' )
            with svg.g(transform=' scale(1 -1) '):
              svg.text(text=((self.y_max-self.y_min)/5*i), x = -40, y = -((y_range/5)*i) + 6, style="color:black") #max y
        for k,s in self.series.items(): #each series
          for p in s:
            print(p)
          svg.polyline((p for p in s), style=f'fill:none;stroke:{colors[color_picker]};stroke-width:1')
          color_picker = (1+color_picker) % 10
          print(color_picker)

      #text
      svg.text(text=self.title, x=self.svg_width/1.9, y = self.label_height*1.4, style=f"color:black;font-size:{self.title_height}") #title
      svg.text(text=self.y_label, x=0, y = self.svg_height/2, style="color:black") #x axis
      svg.text(text=self.x_label, x=self.svg_width/2, y = self.svg_height, style="color:black") #y axis
      svg.text(text=str(self.x_min), x = chart_x_min, y = (chart_y_max+15), style="color:black") #min x
      svg.text(text=str(self.y_min), x = chart_x_min-10, y = (chart_y_max), style="color:black") #min y
      svg.text(text=str(self.x_max), x = chart_x_max-20, y = (chart_y_max+15), style="color:black") #max x
      svg.text(text=str(self.y_max), x = chart_x_min-25, y = (chart_y_min+10), style="color:black") #max y
      #horizontal lines




      pass
