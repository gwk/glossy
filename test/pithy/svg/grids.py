#!/usr/bin/env python3

from pithy.svg import *
from sys import stdout

s = 40

with SvgWriter(stdout, w=1024, h=1024, stroke='black', fill='white', stroke_width=2) as svg:
  y = 5
  svg.grid(title='snap=False', pos=(5, y), size=s, step=10, r=4)
  svg.grid(title='snap=True', pos=(55, y), size=s, step=10, r=4, off=(5, 5))
