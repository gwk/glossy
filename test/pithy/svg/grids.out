<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" stroke="black" fill="white" stroke-width="2" width="1024" height="1024">
<g class="grid">
<title>snap=False</title>
<rect class="grid-border" x="5.0" y="5.0" width="40.0" height="40.0" rx="4" ry="4"/>
<line x1="15.0" y1="5.0" x2="15.0" y2="45.0"/>
<line x1="25.0" y1="5.0" x2="25.0" y2="45.0"/>
<line x1="35.0" y1="5.0" x2="35.0" y2="45.0"/>
<line x1="5.0" y1="15.0" x2="45.0" y2="15.0"/>
<line x1="5.0" y1="25.0" x2="45.0" y2="25.0"/>
<line x1="5.0" y1="35.0" x2="45.0" y2="35.0"/>
</g>
<g class="grid">
<title>snap=True</title>
<rect class="grid-border" x="55.0" y="5.0" width="40.0" height="40.0" rx="4" ry="4"/>
<line x1="60.0" y1="5.0" x2="60.0" y2="45.0"/>
<line x1="70.0" y1="5.0" x2="70.0" y2="45.0"/>
<line x1="80.0" y1="5.0" x2="80.0" y2="45.0"/>
<line x1="90.0" y1="5.0" x2="90.0" y2="45.0"/>
<line x1="55.0" y1="10.0" x2="95.0" y2="10.0"/>
<line x1="55.0" y1="20.0" x2="95.0" y2="20.0"/>
<line x1="55.0" y1="30.0" x2="95.0" y2="30.0"/>
<line x1="55.0" y1="40.0" x2="95.0" y2="40.0"/>
</g>
</svg>
